### PR TITLE
Gutenberg: Replace Gutenlypso with Calypsoified iframe

### PR DESCRIPTION
**Pull request by vindl**
Related P2 posts: p7jreA-26b-p2, p7jreA-24O-p2, p7jreA-25P-p2. 

#### Changes proposed in this Pull Request

* Replace Gutenlypso with Calypsoified iframe for Simple sites. The existing production behavior for Atomic, Jetpack, and VIP sites should remain unchanged.

* In order to simplify the code and flows, I decided to replace the `isGutenbergEnabled` with `isCalypsoifyGutenbergEnabled`. `isGutenbergEnabled` was previously used as a generic way to check both for Gutenlypso (Simple sites) and Calypsoify (Atomic). Since we are removing Gutenlypso now, we can boil this down to just Calypsoify checks. This also allows us to remove the `gutenberg` feature flag completely, which was controlling the Gutenlypso functionality so far.

* After this change, a lot of the Gutenlypso code will become unused. In order to keep this PR from exploding in size and make it easier to review, I intentionally didn't delete it. I'll make sure to clean up the remaining unused code in follow-up PRs.

#### Testing instructions

For Simple sites:

0. The behavior for Atomic, Jetpack, and VIP sites should match production.
1. Apply ~~D23405-code~~ and ~~D23473-code~~ to your sandbox.
2. Use sandboxed Simple site for your testing.
3. Verify that opt-in and opt-out flows still work - try clicking both the buttons available in Inline Help, and `Switch to Classic` in wp-admin.
4. Verify that Inline Help is working as expected.
5. Test various editor flows and make sure that everything still works (publishing, updating). Try with various post types.
6. Make sure that back button in Gutenberg editor is working properly - it should navigate the parent frame, and not in the iframe (note: redirects to `wordpress.com`, might be confusing while testing).

### CODE REVIEW

**Comment by gwwar** [comment]
Fantastic work @vindl ✨! This looks really promising. Overall would you like to treat this as the minimal feature branch to switch this over, or take a feature flag approach ? Let us know in a post.

We'll want to make sure we've covered:
- [ ] Happychat testing
- [ ] Media modal ( @rodrigoi do coordinate with @vindl on folding in your proof of concept  )
- [ ] Calypso revisions (we might as well)
- [ ] Navigation and switching flows (we can open up testing to help break things)

#### Minor Notes:

Lots of extra scrollbars!
screen shot 2019-01-23 at 4 27 53 pm

There's probably other ways of doing this (so feel free to search for alternatives), but I removed the others with:
```css
html { overflow: hidden } //we'll need a better class for this
.main.calypsoify.is-iframe { overflow: hidden };
```
And for good measure I added a scroll="no" attribute to the iframe, because they are finicky.

**Comment by rodrigoi** [comment]
Thanks for taking the time to create this PR. I was a bit overwhelmed about the editor fragmentation and went on a downward spiral of deleting code (mostly from `client/post-editor`). Felt good but led nowhere productive.

When sandboxing a site with a paid plan, we get a _White Screen of Death_ because plugins are a mess.

screen shot 2019-01-23 at 23 38 18

So I took the liberty to update the testing instructions to include a patch for that. Otherwise, this is working very well 😄 

I'll remove the duplicated code from [`try/iframed-gutenberg-with-media-modal`](https://github.com/Automattic/wp-calypso/compare/try/iframed-gutenberg-with-media-modal) and create a new PR just with the `pub/sub` and the WordPress.com Modal overrides.

File: client/gutenberg/editor/calypsoify-iframe.jsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+export default connect( ( state, ownProps ) => {
+	const siteId = getSelectedSiteId( state );
+	const frameNonce = getSiteOption( state, siteId, 'frame_nonce' ) || '';
+	const siteAdminUrl = getSiteAdminUrl( state, siteId );
+
+	const iframeUrl = addQueryArgs(
+		{
+			'frame-nonce': frameNonce,
+		},
+		getIframeUrl( siteAdminUrl, ownProps.postId, ownProps.postType )
```
Comment (rodrigoi) [inline client/gutenberg/editor/calypsoify-iframe.jsx]: We can calculate the iframe url with a bit of lodash and remove `getIframeUrl`

```js
const iframeUrl = addQueryArgs( pickBy(
	{
		'post': postId,
		'action': ( postId && 'edit' ), //if postId, edit.
		'post_type': ( postType !== 'post' && postType ), //if not post, use postType
		'calypsoify': 1,
		'frame-nonce': frameNonce,
	} ),
	getSiteAdminUrl( state, siteId, postId ? 'post.php' : 'post-new.php' )
);
```

has a bit of a cleverness factor, but we can always add some comments ;)

File: client/gutenberg/editor/calypsoify-iframe.jsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	if ( postId ) {
+		return `${ siteAdminUrl }post.php?post=${ postId }&action=edit&calypsoify=1`;
+	}
+	if ( 'post' === postType ) {
+		return `${ siteAdminUrl }post-new.php?calypsoify=1`;
+	}
+	return `${ siteAdminUrl }post-new.php?post_type=${ postType }&calypsoify=1`;
+};
+
+class CalypsoifyIframe extends Component {
```
Comment (rodrigoi) [inline client/gutenberg/editor/calypsoify-iframe.jsx]: for those 👀: we could use a stateless component, but we'll need local state and lifecycle methods down the road to support `pub/sub` with the iframe.

**Comment by mmtr** [comment]
Wonderful @vindl! This works very well!

I noted some differences when comparing it with Gutenlypso. Anyway, I think all of them could be fixed in follow-up PRs if we take a feature flag approach.

**Trash a post**

If you trash a post within the editor, it should redirect to `<CALYPSO_URL>/posts/trashed/<:site>` but instead, the iframe is redirected to `https://<site>.wordpress.com/wp-admin/post-new.php?calypsoify=1` and the browser remains at `<CALYPSO_URL>/block-editor/post/<:site>`.

This causes a not 'sameorigin' error: 
image

**URL is not updated when the draft is saved**

On Gutenlypso, when we start at `/block-editor/post/<:site>`, as soon as we save the post, the URL changes to `/block-editor/post/<:site>/<:post-id>`. This doesn't occur with the Calypsoified iframe.

**Lack of device selector when previewing a post**

One of the benefits of Gutenlypso was that we were able to bring back the support for previewing different screensizes (https://github.com/Automattic/wp-calypso/issues/28477, https://github.com/Automattic/wp-calypso/pull/29321). I believe this was impossible to implement with Calypsoify, but maybe it is possible now with the iframe.

**Site title is missing on the header**
_Gutenlypso_
screen shot 2019-01-24 at 16 18 16

_Calypsoified iframe_
screen shot 2019-01-24 at 16 18 08

**Comment by vindl** [comment]
> Overall would you like to treat this as the minimal feature branch to switch this over, or take a feature flag approach? Let us know in a post.

My plan was to use this as a feature branch and avoid adding a new feature flag for the switch. The situation was complicated enough with existing 3 feature flags, and adding a new one would just make things worse. 😄I'll write more details in the upcoming post.

**Comment by gwwar** [comment]
Nice finds @mmtr @rodrigoi 💖

**Comment by vindl** [comment]
> Lots of extra scrollbars!

@gwwar I had trouble reproducing this (can only see the content scrollbar) - what browser were you using?

File: client/gutenberg/editor/calypsoify-iframe.jsx (hunk 1 above)
Comment (vindl) [reply client/gutenberg/editor/calypsoify-iframe.jsx]: Thanks for the suggestion! Your approach looks better to me, refactored in 0b836f55e535d2523235105beeb442022be9987f.

**Comment by vindl** [comment]
>URL is not updated when the draft is saved
>
>On Gutenlypso, when we start at /block-editor/post/<:site>, as soon as we save the post, the URL >changes to /block-editor/post/<:site>/<:post-id>. This doesn't occur with the Calypsoified iframe.

I think we'll need to reuse some of the `postMessage` logic that @rodrigoi's Media Modal PR brings in order to communicate this. I can't figure out any other way to resolve it.

> Lack of device selector when previewing a post
> 
> One of the benefits of Gutenlypso was that we were able to bring back the support for previewing different screensizes (#28477, #29321). I believe this was impossible to implement with Calypsoify, but maybe it is possible now with the iframe.

I think it should be possible now too, but it's not a blocker for the first release of Calypsoify iframe (see discussion in p7jreA-26b-p2 for more details).

**Comment by gwwar** [comment]
> @gwwar I had trouble reproducing this (can only see the content scrollbar) - what browser were you using?

Chrome, on OSX. OS UI is setup to not hide scrollbars by default.

**Comment by bsessions85** [comment]
I am working on updating the e2e tests to work with this. I noticed that when I click the view page link after publishing, because we're in an iframe, the url shown doesn't change. If someone were to copy that url thinking that is the url of their post, it wouldn't be correct

**Comment by gwwar** [comment]
>  If someone were to copy that url thinking that is the url of their post, it wouldn't be correct

Right @bsessions85, this is what @vindl is referring to in https://github.com/Automattic/wp-calypso/pull/30307#issuecomment-458126416

We'll need some iframe messaging to reflect parent url updates.

**Comment by gwwar** [comment]
closing in favor of https://github.com/Automattic/wp-calypso/pull/30451

**Pull request closed without merging**